jetty-jmh/src/main/java/org/eclipse/jetty/io/jmh/ByteBufferBenchmark.java

```diff
@@ -0,0 +1,154 @@
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+
+@State(Scope.Benchmark)
+@Threads(4)
+@Warmup(iterations = 7, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 7, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+public class ByteBufferBenchmark
+{
+    public long test(ByteBuffer buffer)
+    {
+        buffer.clear();
+        while(buffer.hasRemaining())
+        {
+            int size = ThreadLocalRandom.current().nextInt(1024);
+            byte[] bytes = new byte[size];
+            ThreadLocalRandom.current().nextBytes(bytes);
+            buffer.put(bytes,0,Math.min(bytes.length,buffer.remaining()));
+        }
+
+        buffer.flip();
+
+        long sum = 0;
+        while(buffer.hasRemaining())
+            sum += buffer.get();
+
+        return sum;
+    }
+
+
+    public long testArray(ByteBuffer buffer)
+    {
+        buffer.clear();
+        byte[] array = buffer.array();
+        int offset = buffer.arrayOffset();
+        int end = offset + buffer.remaining();
+        while(offset<end)
+        {
+            int size = ThreadLocalRandom.current().nextInt(1024);
+            byte[] bytes = new byte[size];
+            ThreadLocalRandom.current().nextBytes(bytes);
+            System.arraycopy(bytes,0,array,offset,Math.min(bytes.length,end-offset));
+            offset += bytes.length;
+        }
+        buffer.position(buffer.limit());
+        buffer.flip();
+
+        long sum = 0;
+        array = buffer.array();
+        offset = buffer.arrayOffset();
+        end = offset + buffer.remaining();
+
+        while(offset<end)
+            sum += array[offset++];
+        buffer.position(buffer.limit());
+        return sum;
+    }
+
+
+
+    @Benchmark
+    @BenchmarkMode({ Mode.Throughput})
+    public long testDirect()
+    {
+        ByteBuffer buffer = ByteBuffer.allocateDirect(32768);
+        long sum = 0;
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        return sum;
+    }
+
+
+    @Benchmark
+    @BenchmarkMode({ Mode.Throughput})
+    public long testInDirect()
+    {
+        ByteBuffer buffer = ByteBuffer.allocate(32768);
+        long sum = 0;
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        sum ^= test(buffer);
+        return sum;
+    }
+
+    @Benchmark
+    @BenchmarkMode({ Mode.Throughput})
+    public long testInDirectArray()
+    {
+        ByteBuffer buffer = ByteBuffer.allocate(32768);
+        long sum = 0;
+        sum ^= testArray(buffer);
+        sum ^= testArray(buffer);
+        sum ^= testArray(buffer);
+        sum ^= testArray(buffer);
+        sum ^= testArray(buffer);
+        return sum;
+    }
```

Comment:
I had to double check that all benchmark methods actually called the `test*()` method 5 times.

It would be better to use a `java.util.function.ToLongFunction<ByteBuffer>` from the benchmark methods and call a single method that takes the `ToLongFunction` as parameter and invokes it 5 times, so that all benchmark methods call the same method (with slightly different `ToLongFunction`s) and there is no doubt about how many times the `test*()` methods is called by each benchmark - and remove duplication.